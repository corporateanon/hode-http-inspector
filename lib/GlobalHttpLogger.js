const http = require('http');
const https = require('https');
const nodeMajorVersion = +process.version.match(/^v(\d+)\.(\d+)\.(\d+)$/)[1];

class ResponseBodyCollector {
    constructor(response) {
        this.buffers = [];
        this.bodyPromise = new Promise((resolve, reject) => {
            response.on('data', chunk => {
                this.buffers.push(chunk);
            });
            response.on('end', () => {
                const body = Buffer.concat(this.buffers);
                resolve(body);
            });
        });
    }
    getBodyAsync() {
        return this.bodyPromise;
    }
}

const waitForResponseOrError = request =>
    new Promise((resolve, reject) => {
        request.prependOnceListener('response', response => {
            response.bodyCollector = new ResponseBodyCollector(response);
            resolve({ response });
        });
        request.prependOnceListener('error', error => resolve({ error }));
    });

const collectRequestBody = request =>
    new Promise((resolve, reject) => {
        const requestBody = [];

        const reqWrite = request.write;
        request.write = function() {
            /**
             * chunk can be either a string or a Buffer.
             */
            const chunk = arguments[0];

            if (Buffer.isBuffer(chunk)) {
                requestBody.push(chunk);
            } else {
                requestBody.push(Buffer.from(chunk, 'utf8'));
            }

            return reqWrite.apply(this, arguments);
        };

        const reqEnd = request.end;
        request.end = function() {
            /**
             * the first argument might be a callback or a chunk
             */
            const maybeChunk = arguments[0];

            if (Buffer.isBuffer(maybeChunk)) {
                requestBody.push(maybeChunk);
            } else if (maybeChunk && typeof maybeChunk !== 'function') {
                requestBody.push(Buffer.from(maybeChunk, 'utf8'));
            }

            return reqEnd.apply(this, arguments);
        };

        request.prependOnceListener('finish', () => {
            if (!requestBody.length) {
                resolve(null);
            } else {
                resolve(Buffer.concat(requestBody));
            }
        });
    });

const interceptRequest = async (request, { onRequestEnd }) => {
    const requestBody = await collectRequestBody(request);
    let responseBody = null;

    const { response, error } = await waitForResponseOrError(request);
    if (response) {
        responseBody = await response.bodyCollector.getBodyAsync();
    } else if (error) {
        debugger;
    }

    const protocol = request.agent.protocol;
    const host = request.getHeader('host');
    const path = request.path;

    onRequestEnd({
        request: {
            url: `${protocol}//${host}${path}`,
            method: request.method,
            headers: request._headers,
            body: requestBody
        },
        response: {
            status: response.statusCode,
            body: responseBody,
            headers: response.headers
        }
    });
};

module.exports = class GlobalHttpLogger {
    constructor({ onRequestEnd }) {
        this.onRequestEnd = onRequestEnd;
    }
    start() {
        const { onRequestEnd } = this;
        const interceptedRequestMethod = (object, func, ...rest) => {
            const req = func.call(object, ...rest);
            interceptRequest(req, { onRequestEnd });
            return req;
        };

        http.request = interceptedRequestMethod.bind(null, http, http.request);
        http.get = interceptedRequestMethod.bind(null, http, http.get);

        /**
         * https.request proxies to http.request for 8.x and earlier versions
         */
        if (nodeMajorVersion > 8) {
            https.get = interceptedRequestMethod.bind(null, https, https.get);
            https.request = interceptedRequestMethod.bind(
                null,
                https,
                https.request
            );
        }
    }
    stop() {}
};
