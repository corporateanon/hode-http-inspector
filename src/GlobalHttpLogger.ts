// https://github.com/Microsoft/TypeScript-Babel-Starter

import http, { ClientRequest, IncomingMessage } from 'http';
import https from 'https';
import {
    LoggerEvent,
    LoggerHeaders,
    LoggerRequest,
    LoggerResponse,
    LoggerEventHandler,
    LoggerShouldLog,
    Timestamp,
} from './SharedTypes';
import { createUnzip, Unzip } from 'zlib';
import nanoid from 'nanoid';

const matches = process.version.match(/^v(\d+)\.(\d+)\.(\d+)$/);
const nodeMajorVersion = matches ? +matches[1] : 0;

interface ClientRequestWithUndocumentedMembers extends ClientRequest {
    agent: any;
    method: string;
    _headers: { [headerName: string]: string };
}

class ResponseBodyCollector {
    buffers: Array<Buffer>;

    bodyPromise: Promise<Buffer>;

    private shouldUnzip(response: IncomingMessage): boolean {
        const encoding = response.headers['content-encoding'];
        const status = response.statusCode;
        return (
            !!encoding &&
            ['gzip', 'deflate', 'compress'].includes(encoding) &&
            status !== 204
        );
    }

    private unzip(response: IncomingMessage): Unzip {
        const stream = response.pipe(createUnzip());
        return stream;
    }

    constructor(response: IncomingMessage) {
        this.buffers = [];

        const stream = this.shouldUnzip(response)
            ? this.unzip(response)
            : response;

        this.bodyPromise = new Promise((resolve) => {
            stream.on('data', (chunk) => {
                this.buffers.push(chunk);
            });
            stream.on('end', () => {
                const body = Buffer.concat(this.buffers);
                resolve(body);
            });
        });
    }

    getBodyAsync(): Promise<Buffer> {
        return this.bodyPromise;
    }
}

const waitForResponseOrError = (
    request: ClientRequest
): Promise<{
    response?: IncomingMessage;
    responseBodyCollector?: ResponseBodyCollector;
    responseTimeStart?: Timestamp;
    error?: Error;
}> =>
    new Promise((resolve) => {
        request.prependOnceListener('response', (response) => {
            const responseTimeStart = Date.now();
            const responseBodyCollector = new ResponseBodyCollector(response);
            resolve({ response, responseBodyCollector, responseTimeStart });
        });
        request.prependOnceListener('error', (error) => {
            resolve({ error });
        });
    });

const collectRequestBody = (request: ClientRequest): Promise<Buffer | null> =>
    new Promise((resolve) => {
        const requestBody: Array<Buffer> = [];

        const reqWrite = request.write;
        request.write = function (...args: any) {
            /**
             * chunk can be either a string or a Buffer.
             */
            const chunk = args[0];

            if (Buffer.isBuffer(chunk)) {
                requestBody.push(chunk);
            } else {
                requestBody.push(Buffer.from(chunk, 'utf8'));
            }

            return reqWrite.apply(this, args);
        };

        const reqEnd = request.end;
        request.end = function (...args: any) {
            /**
             * the first argument might be a callback or a chunk
             */
            const maybeChunk = args[0];

            if (Buffer.isBuffer(maybeChunk)) {
                requestBody.push(maybeChunk);
            } else if (maybeChunk && typeof maybeChunk !== 'function') {
                requestBody.push(Buffer.from(maybeChunk, 'utf8'));
            }

            return reqEnd.apply(this, args);
        };

        request.prependOnceListener('finish', () => {
            if (!requestBody.length) {
                resolve(null);
            } else {
                resolve(Buffer.concat(requestBody));
            }
        });

        request.prependOnceListener('error', () => resolve(null));
    });

const interceptRequest = async (
    request: ClientRequest,
    onRequestEnd: (payload: LoggerEvent) => void,
    shouldLog?: LoggerShouldLog
) => {
    const protocol = (request as ClientRequestWithUndocumentedMembers).agent
        .protocol;
    const host = request.getHeader('host');
    const path = request.path;

    const loggerRequest: LoggerRequest = {
        id: nanoid(),
        timeStart: Date.now(),
        url: `${protocol}//${host}${path}`,
        method: (request as ClientRequestWithUndocumentedMembers).method,
        headers: (request as ClientRequestWithUndocumentedMembers)._headers,
        body: null,
    };

    if (shouldLog && !shouldLog(loggerRequest)) {
        return;
    }

    const [
        requestBody,
        { response, responseBody, responseTimeStart, error },
    ] = await Promise.all([
        collectRequestBody(request),
        (async () => {
            const {
                response,
                responseBodyCollector,
                responseTimeStart,
                error,
            } = await waitForResponseOrError(request);
            if (response && responseBodyCollector) {
                return {
                    response,
                    responseTimeStart,
                    responseBody: await responseBodyCollector.getBodyAsync(),
                    error: null,
                };
            } else if (error) {
                return {
                    response: null,
                    responseBody: null,
                    error,
                };
            } else {
                throw new Error('No responseBodyCollector');
            }
        })(),
    ]);

    loggerRequest.body = requestBody ? requestBody : null;

    if (response) {
        const loggerResponse: LoggerResponse = {
            timeStart: responseTimeStart ?? 0,
            status: response.statusCode ?? 0,
            body: responseBody ? responseBody : null,
            headers: response.headers as LoggerHeaders,
        };
        onRequestEnd({
            request: loggerRequest,
            response: loggerResponse,
            error: null,
        });
    } else if (error) {
        const loggerError = {
            message: error.message,
            code: (error as any).code as string,
            stack: error.stack ?? '',
        };
        onRequestEnd({
            request: loggerRequest,
            response: null,
            error: loggerError,
        });
    }
};

export interface GlobalHttpLoggerOptions {
    onRequestEnd: LoggerEventHandler;
    shouldLog?: LoggerShouldLog;
}

export default class GlobalHttpLogger {
    private onRequestEnd: LoggerEventHandler;
    private shouldLog?: LoggerShouldLog;
    constructor({ onRequestEnd, shouldLog }: GlobalHttpLoggerOptions) {
        this.onRequestEnd = onRequestEnd;
        this.shouldLog = shouldLog;
    }
    start(): void {
        const { onRequestEnd } = this;
        const interceptedRequestMethod = (
            object: any,
            func: any,
            ...rest: any
        ) => {
            const req = func.call(object, ...rest);
            interceptRequest(req, onRequestEnd, this.shouldLog);
            return req;
        };

        http.request = interceptedRequestMethod.bind(null, http, http.request);
        http.get = interceptedRequestMethod.bind(null, http, http.get);

        /**
         * https.request proxies to http.request for 8.x and earlier versions
         */
        if (nodeMajorVersion > 8) {
            https.get = interceptedRequestMethod.bind(null, https, https.get);
            https.request = interceptedRequestMethod.bind(
                null,
                https,
                https.request
            );
        }
    }
}
