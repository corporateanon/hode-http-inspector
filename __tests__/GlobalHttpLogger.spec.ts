import axios from 'axios';
import sleep from 'sleep-promise';
import GlobalHttpLogger from '../src/GlobalHttpLogger';
import { prepareSnapshot } from './utils';

const skipTick = () => sleep(0);

describe('GlobalHttpLogger', () => {
    it('should not capture requests if shouldLog returns false', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({
            onRequestEnd,
            shouldLog: (req) => !req.headers['x-http-inspector-do-not-track'],
        });
        logger.start();
        const res = await axios.get('http://127.0.0.1:13000/get?a=b', {
            headers: {
                'x-http-inspector-do-not-track': '1',
            },
        });
        expect(res.data).toEqual('Hello World!');
        expect(res.status).toEqual(200);
        await skipTick();
        expect(onRequestEnd.mock.calls.length).toEqual(0);
    });
    it('should capture GET requests', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({ onRequestEnd });
        logger.start();
        const res = await axios.get('http://127.0.0.1:13000/get?a=b');
        expect(res.data).toEqual('Hello World!');
        expect(res.status).toEqual(200);
        await skipTick();
        expect(
            prepareSnapshot(onRequestEnd.mock.calls[0][0])
        ).toMatchSnapshot();
    });
    it('should capture POST requests', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({ onRequestEnd });
        logger.start();
        const res = await axios.post('http://127.0.0.1:13000/post', {
            foo: 'bar',
        });
        expect(res.data).toEqual({ hello: 'world' });
        expect(res.status).toEqual(200);
        await skipTick();
        expect(
            prepareSnapshot(onRequestEnd.mock.calls[0][0])
        ).toMatchSnapshot();
    });
    it('should capture GET requests with error response', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({ onRequestEnd });
        logger.start();
        const res = await axios.get('http://127.0.0.1:13000/get/404', {
            validateStatus: () => true,
        });
        expect(res.data).toEqual({ error: 'Not found' });
        expect(res.status).toEqual(404);
        await skipTick();
        expect(
            prepareSnapshot(onRequestEnd.mock.calls[0][0])
        ).toMatchSnapshot();
    });
    it('should capture GET requests with network error', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({ onRequestEnd });
        logger.start();
        await expect(
            axios.get('http://never.existing.host.asdfgh/', {
                validateStatus: () => true,
            })
        ).rejects.toThrow('getaddrinfo ENOTFOUND never.existing.host.asdfgh');
        await skipTick();
        const event = prepareSnapshot(onRequestEnd.mock.calls[0][0]);
        expect(event.request).toMatchSnapshot();
        expect(event.response).toEqual(null);
        expect(event.error?.code).toEqual('ENOTFOUND');
        expect([
            'getaddrinfo ENOTFOUND never.existing.host.asdfgh',
            'getaddrinfo ENOTFOUND never.existing.host.asdfgh never.existing.host.asdfgh:80',
        ]).toContain(event.error?.message);
    });

    it('should decode gzipped content', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({ onRequestEnd });
        logger.start();
        const res = await axios.get('http://127.0.0.1:13000/compressable', {
            headers: {
                'accept-encoding': 'gzip',
            },
        });
        expect(res.data).toEqual('Hello World!');
        expect(res.status).toEqual(200);
        await skipTick();
        expect(
            prepareSnapshot(onRequestEnd.mock.calls[0][0])
        ).toMatchSnapshot();
    });

    it('should allow multiple headers', async () => {
        const onRequestEnd = jest.fn();
        const logger = new GlobalHttpLogger({ onRequestEnd });
        logger.start();
        const res = await axios.get('http://127.0.0.1:13000/get?a=b', {
            headers: {
                one: ['two', 'three'],
            },
        });
        expect(res.data).toEqual('Hello World!');
        expect(res.status).toEqual(200);
        await skipTick();
        expect(
            prepareSnapshot(onRequestEnd.mock.calls[0][0])
        ).toMatchSnapshot();
    });
});
