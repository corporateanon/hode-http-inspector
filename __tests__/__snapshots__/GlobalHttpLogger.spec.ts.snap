// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalHttpLogger should allow multiple headers 1`] = `
Object {
  "error": null,
  "request": Object {
    "body": null,
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "host": "127.0.0.1:13000",
      "one": Object {
        "0": "two",
        "1": "three",
      },
      "user-agent": "axios/0.19.2",
    },
    "id": "[presents]",
    "method": "GET",
    "timeStart": 1,
    "url": "http://127.0.0.1:13000/get?a=b",
  },
  "response": Object {
    "body": "Hello World!",
    "headers": Object {
      "connection": "close",
      "content-length": "12",
      "content-type": "text/html; charset=utf-8",
      "vary": "Accept-Encoding",
      "x-powered-by": "Express",
    },
    "rawHeaders": Array [
      "X-Powered-By",
      "Express",
      "Content-Type",
      "text/html; charset=utf-8",
      "Content-Length",
      "12",
      "Vary",
      "Accept-Encoding",
      "Connection",
      "close",
    ],
    "status": 200,
    "timeStart": 1,
  },
}
`;

exports[`GlobalHttpLogger should capture GET requests 1`] = `
Object {
  "error": null,
  "request": Object {
    "body": null,
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "host": "127.0.0.1:13000",
      "user-agent": "axios/0.19.2",
    },
    "id": "[presents]",
    "method": "GET",
    "timeStart": 1,
    "url": "http://127.0.0.1:13000/get?a=b",
  },
  "response": Object {
    "body": "Hello World!",
    "headers": Object {
      "connection": "close",
      "content-length": "12",
      "content-type": "text/html; charset=utf-8",
      "vary": "Accept-Encoding",
      "x-powered-by": "Express",
    },
    "rawHeaders": Array [
      "X-Powered-By",
      "Express",
      "Content-Type",
      "text/html; charset=utf-8",
      "Content-Length",
      "12",
      "Vary",
      "Accept-Encoding",
      "Connection",
      "close",
    ],
    "status": 200,
    "timeStart": 1,
  },
}
`;

exports[`GlobalHttpLogger should capture GET requests with error response 1`] = `
Object {
  "error": null,
  "request": Object {
    "body": null,
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "host": "127.0.0.1:13000",
      "user-agent": "axios/0.19.2",
    },
    "id": "[presents]",
    "method": "GET",
    "timeStart": 1,
    "url": "http://127.0.0.1:13000/get/404",
  },
  "response": Object {
    "body": "{\\"error\\":\\"Not found\\"}",
    "headers": Object {
      "connection": "close",
      "content-length": "21",
      "content-type": "application/json; charset=utf-8",
      "vary": "Accept-Encoding",
      "x-powered-by": "Express",
    },
    "rawHeaders": Array [
      "X-Powered-By",
      "Express",
      "Content-Type",
      "application/json; charset=utf-8",
      "Content-Length",
      "21",
      "Vary",
      "Accept-Encoding",
      "Connection",
      "close",
    ],
    "status": 404,
    "timeStart": 1,
  },
}
`;

exports[`GlobalHttpLogger should capture GET requests with network error 1`] = `
Object {
  "body": null,
  "headers": Object {
    "accept": "application/json, text/plain, */*",
    "host": "never.existing.host.asdfgh",
    "user-agent": "axios/0.19.2",
  },
  "id": "[presents]",
  "method": "GET",
  "timeStart": 1,
  "url": "http://never.existing.host.asdfgh/",
}
`;

exports[`GlobalHttpLogger should capture POST requests 1`] = `
Object {
  "error": null,
  "request": Object {
    "body": "{\\"foo\\":\\"bar\\"}",
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "content-length": 13,
      "content-type": "application/json;charset=utf-8",
      "host": "127.0.0.1:13000",
      "user-agent": "axios/0.19.2",
    },
    "id": "[presents]",
    "method": "POST",
    "timeStart": 1,
    "url": "http://127.0.0.1:13000/post",
  },
  "response": Object {
    "body": "{\\"hello\\":\\"world\\"}",
    "headers": Object {
      "connection": "close",
      "content-length": "17",
      "content-type": "application/json; charset=utf-8",
      "vary": "Accept-Encoding",
      "x-powered-by": "Express",
    },
    "rawHeaders": Array [
      "X-Powered-By",
      "Express",
      "Content-Type",
      "application/json; charset=utf-8",
      "Content-Length",
      "17",
      "Vary",
      "Accept-Encoding",
      "Connection",
      "close",
    ],
    "status": 200,
    "timeStart": 1,
  },
}
`;

exports[`GlobalHttpLogger should decode gzipped content 1`] = `
Object {
  "error": null,
  "request": Object {
    "body": null,
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "accept-encoding": "gzip",
      "host": "127.0.0.1:13000",
      "user-agent": "axios/0.19.2",
    },
    "id": "[presents]",
    "method": "GET",
    "timeStart": 1,
    "url": "http://127.0.0.1:13000/compressable",
  },
  "response": Object {
    "body": "Hello World!",
    "headers": Object {
      "connection": "close",
      "content-type": "text/html; charset=utf-8",
      "transfer-encoding": "chunked",
      "vary": "Accept-Encoding",
      "x-powered-by": "Express",
    },
    "rawHeaders": Array [
      "X-Powered-By",
      "Express",
      "Content-Type",
      "text/html; charset=utf-8",
      "Vary",
      "Accept-Encoding",
      "Content-Encoding",
      "gzip",
      "Connection",
      "close",
      "Transfer-Encoding",
      "chunked",
    ],
    "status": 200,
    "timeStart": 1,
  },
}
`;
